cfblog/models.py

```diff
@@ -0,0 +1,146 @@
+from __future__ import unicode_literals
+
+__author__ = 'vinay'
+
+import datetime
+
+from django.conf import settings
+from django.db import models
+from django.core.exceptions import ValidationError
+from django.core.urlresolvers import resolve
+from django.utils.translation import ugettext_lazy as _
+
+from jsonfield import JSONField
+from tagging.fields import TagField
+
+from .managers import CmsBlogCategoryManager, CmsBlogPostManager
+from .utils import cacheable, stale_cache, get_public_data, get_image_path, get_file_path
+from .validators import validate_and_get_template, validate_url_path
+
+
+class CmsPageTemplate(models.Model):
+    name = models.CharField(_('template name'), max_length=255, db_index=True, validators=[validate_and_get_template])
+
+    def __unicode__(self):
+        return self.name
+
+
+STATUS_CHOICES = (
+    (1, _('Draft')),
+    (2, _('Public')),
+)
+
+
+class BaseCmsPage(models.Model):
+    template = models.ForeignKey(CmsPageTemplate, verbose_name=_('template'), null=True, blank=True)
+    auth_data = JSONField(verbose_name=_('author data'), default={}, blank=True)
+    public_data = JSONField(verbose_name=_('public data'), default={}, blank=True)
+    created_on = models.DateTimeField(_('created on'), auto_now_add=True)
+    modified_on = models.DateTimeField(_('modified on'), auto_now=True)
+    status = models.IntegerField(_('status'), choices=STATUS_CHOICES, default=1)
+    publish = models.DateTimeField(_('publish'), default=datetime.datetime.now)
+
+    class Meta:
+        abstract = True
+
+    @property
+    def is_public(self):
+        return self.status == 2
+
+    def publish_cms_content(self):
+        self.public_data = self.auth_data
+        self.status = 2
+        self.save()
+
+
+class CmsPage(BaseCmsPage):
+    url = models.CharField(_('url path'),
+                           max_length=255, db_index=True, unique=True, validators=[validate_url_path])
+
+    @cacheable('cms_page_{id}')
+    def html(self):
+        return get_public_data(self)
+
+    @stale_cache('cms_page_{id}')
+    def save(self, *args, **kwargs):
+        super(CmsPage, self).save(*args, **kwargs)
+
+    def clean(self):
+        from .views import cms_page_index
+        resolver_math = resolve(self.url)
+        if resolver_math.func is cms_page_index and not self.template:
+            raise ValidationError(_(
+                'Template cannot be empty for new urls'
+            ))
+
+        if resolver_math.func is not cms_page_index and self.template:
+            raise ValidationError(_(
+                'Template is not required for existing urls'
+            ))
+        super(CmsPage, self).clean()
+
+
+class CmsBlogCategory(models.Model):
```

Comment:
This is over designed. We have simple non nested category. Will have only one fields title, description and slug. Remove level, is_active, and parent.